metadata-editor/app/controllers/EditsController.scala

```diff
@@ -39,12 +43,43 @@ import scala.concurrent.{ExecutionContext, Future}
 //   }
 // }
 
-class EditsController(auth: Authentication, store: EditsStore, notifications: Notifications, config: EditsConfig,
-                      override val controllerComponents: ControllerComponents)(implicit val ec: ExecutionContext)
+class EditsController(
+                       auth: Authentication,
+                       store: EditsStore,
+                       notifications: Notifications,
+                       config: EditsConfig,
+                       ws: WSClient,
+                       authorisation: Authorisation,
+                       override val controllerComponents: ControllerComponents
+                     )(implicit val ec: ExecutionContext)
   extends BaseController with ArgoHelpers with EditsResponse {
 
   import UsageRightsMetadataMapper.usageRightsToMetadata
 
+  val services: Services = new Services(config.domainRoot, ServiceHosts.guardianPrefixes, Set.empty)
+  val gridClient: GridClient = GridClient(services)(ws)
+
+  def AuthorisedToEditMetadataOrUploader(imageId: String): ActionFilter[Request] = new ActionFilter[Request] {
+    override protected def filter[A](request: Request[A]): Future[Option[Result]] = {
+      //We first check for permissions, if the user has permissions we avoid making a call to media-api service
+      val hasPermission: Boolean = authorisation.hasPermissionTo(EditMetadata)(request.user)
+      if (hasPermission) {
+        Future.successful(None)
+      } else {
+        val result = for {
+          uploadedBy <- gridClient.getUploadedBy(imageId, auth.getOnBehalfOfPrincipal(request.user))
+          isAuthorised = authorisation.canUserWriteMetadata(request.user, uploadedBy.getOrElse(""))
+          if isAuthorised
+        } yield {
+          Future.successful(None)
+        }
+        result.flatten.recover{case _ => Some(authorisation.unauthorized(EditMetadata))}
+      }
+    }
+    override protected def executionContext: ExecutionContext = ec
+  }
```

Comment:
This is really nice. 💯 

Without too much work I imagine that this could be made generic and moved into the Authorisation class. Perhaps an `actionFilterForUploaderOr` method? Either the method or `Authorisation` class with need a grid client.

Reply:
I didn't see the use of making this generic, as the only other permission where the uploader of the image matters is DeleteImage, and the deletion of images is done directly in media-api service, so no need to do an external call. However, I may be missing some use case for that, and I'm happy to make this generic if you prefer.

Reply:
I added the refactor as a new commit, showing how it could be used for DeleteImage. However, I didn't use it for that endpoint, since that one already retrieves the whole image to check for syndication rights, so it seemed kind of wasteful to do another query to elasticSearch to get the uploader of the image.